### PR TITLE
v1 of the streamSignupManager Cog

### LINKED ISSUE #80: Add commands to handle stream signups
Need new commands to:
- Apply for a spot on stream for a given match day and time
- Check which teams have applied for a given match day and time
- Remove an application (restricted to just your own team's application)

Applications will go through the following process:
1. Created using the command, saved with the member id of the person who used the command and the team they represent. Marked as "Waiting Verification" on command to see applications.
2. A dm gets sent to the opposing team's captain asking them to verify the application.
3. If opposing team captain rejects the application, delete application and send a dm to the original applicant stating it was rejected.
4. If opposing team captain verifies the application the opposing team captain and his team are then saved to the applicant data as well. Marked as "Verified" on command to see applications.

This may be best written in it's own cog, but will need to use info from the match cog so may also be an option for it to be written as part of that.

Capture

### PR DESCRIPTION
**Pull request by BenjaminHunt**
Added a full set of commands and features for scheduling matches on stream.
- Player initiates application
- Opposing team captain may accept/reject application
- Upon accepting the application, Media Committee is notified if a stream channels feed has been set.

Media Committee may
- Accept/Reject application
- Bypass application process and set any match to be on stream
- Rescind application -- remove a scheduled match from the stream
- DM updates are sent to players at appropriate application status intervals

README.md has been written for a fully comprehensive explanation of commands and behavioral flow.

Match cog has been updated:
- to make getting match information easier for the streamSignupManager cog.
- to update information when a match is scheduled on or removed from stream
- so that the `<p>match` command can reflect more appropriate match information for matches scheduled on stream

resolves #80

### CODE REVIEW

File: match/match.py [hunk 1]
```diff
@@ -150,7 +150,7 @@ async def addMatches(self, ctx, *matches):
 
         Examples:
 
-        [p]addMatches "['1','September 10, 2018','Fire Ants','Leopards',
+        [p]addMatches "['1','September 10, 2020','Fakes','Leopards',
```
Comment (adammast) [inline match/match.py:153]: Is there a reason this was changed? I like having the example show an example of a team name with multiple words to show how that can be handled inside the quotes.

File: match/match.py [hunk 2]
```diff
@@ -278,6 +279,16 @@ async def _add_match(self, ctx, match_day, match_date, home, away, *args):
         result['away'] = away
         return result
 
+    async def _set_match_on_stream(self, ctx, match_day, team, stream_details):
+        matches = await self._matches(ctx)
+        for match in matches:
+            if match['matchDay'] == match_day and (one_team == match['home'] or one_team == match['away']):
```
Comment (adammast) [inline match/match.py:285]: This loop functionality looks the same as in the get_match_from_day_team method. Is there a reason not to just call that method to get the match and edit it here if found? That way the functionality can be guaranteed to be consistent if it needs to change in the future.

File: match/match.py [hunk 3]
```diff
@@ -371,16 +391,77 @@ async def _format_match_message(self, ctx, match_index, user_team_name=None):
...
+        # }
+        for match in matches:
+            if match['matchDay'] == match_day:
+                if match['home'].casefold() == team_name.casefold() or match['away'].casefold() == team_name.casefold():
+                    return match
+        return None
+
+    async def set_match_on_stream(self, ctx, match_day, team_name, stream_info):
+        matches = await self._matches(ctx)
+        for match in matches:
```
Comment (adammast) [inline match/match.py:421]: This loop functionality looks the same as in the get_match_from_day_team method. Is there a reason not to just call that method to get the match and edit it here if found? That way the functionality can be guaranteed to be consistent if it needs to change in the future.

File: match/match.py [hunk 4]
```diff
@@ -371,16 +391,77 @@ async def _format_match_message(self, ctx, match_index, user_team_name=None):
...
+                break 
+            if match['home'] == team_name or match['away'] == team_name:
+                match['streamDetails'] = stream_info
+                await self._save_matches(ctx, matches)
+                return True
+        return False
+
+    async def remove_match_from_stream(self, ctx, match_day, team_name):
+        matches = await self._matches(ctx)
+        for match in matches:
```
Comment (adammast) [inline match/match.py:432]: This loop functionality looks the same as in the get_match_from_day_team method. Is there a reason not to just call that method to get the match and edit it here if found? That way the functionality can be guaranteed to be consistent if it needs to change in the future.

File: streamSignupManager/README.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
+# RSCBot: combineRooms
```
Comment (adammast) [inline streamSignupManager/README.md:1]: This should be streamSignupManager instead of combineRooms, correct?

File: match/match.py (hunk 1 above)
Comment (BenjaminHunt) [reply match/match.py:153]: Nah I was just playing around with it. I'm happy to change it back.

File: streamSignupManager/README.md (hunk 5 above)
Comment (BenjaminHunt) [reply streamSignupManager/README.md:1]: yes, thank you!

File: streamSignupManager/streamSignupManager.py [hunk 6]
```diff
@@ -0,0 +1,1089 @@
...
+        # Application statuses
+        self.PENDING_OPP_CONFIRMATION_STATUS = "PENDING_OPP_CONFIRMATION"
+        self.PENDING_LEAGUE_APPROVAL_STATUS = "PENDING_LEAGUE_APPROVAL"
+        self.SCHEDULED_ON_STREAM_STATUS = "SCHEDULED_ON_STREAM"
+        self.REJECTED_STATUS = "REJECTED"
+        self.RESCINDED_STATUS = "RESCINDED"
+    
+    @commands.command(aliases=["streamapp", "streamapply", "streamApplications", "streamapplications"])
+    @commands.guild_only()
+    async def streamApp(self, ctx, action, match_day, time_slot=None, team=None):
```
Comment (adammast) [inline streamSignupManager/streamSignupManager.py:36]: For the action parameter I think a lot of the code around checking which action it is can be streamlined by using subcommands. That is a built in way to allow multi-word commands in order to be more specific with what the user is trying to do with the command. Look at the sixMans cog, specifically the queueLeaderBoard command and its subcommands such as overall, day, week, etc. Let me know if you have any questions about that.

Comment (BenjaminHunt) [reply streamSignupManager/streamSignupManager.py]: Done.

File: match/match.py (hunk 3 above)
Comment (BenjaminHunt) [reply match/match.py:422]: So this is very similar, but `_get_match_from_day_team` is an accessor command. In this area, I am not just getting the match, but updating it as well. Is there a good way I can use that command as the first step to updating it? The only thing that comes to my mind is if the get_match command either returns the index, or if I make a second command called `_update_match_` or something of the sort.

Comment (BenjaminHunt) [reply match/match.py:422]: So this is very similar, but _get_match_from_day_team is an accessor command. In this area, I am not just getting the match, but updating it as well. Is there a good way I can use that command as the first step to updating it? The only thing that comes to my mind is if the get_match command either returns the index, or if I make a second command called _update_match_ or something of the sort.

What do you think?

Comment (BenjaminHunt) [reply match/match.py:422]: So this is very similar, but `_get_match_from_day_team` is an accessor command. In this area, I am not just getting the match, but updating it as well. Is there a good way I can use that command as the first step to updating it? The only thing that comes to my mind is if the get match command either returns the index, or if I make a second command called `update_match` or something of the sort.

What do you think?

**Review by adammast — APPROVED** [review]

**Pull request merged**